services/api/src/models/group.ts

```diff
@@ -833,6 +834,42 @@ export const Group = (clients: {
     }
   };
 
+  // helper to remove user from groups
+  const removeUserFromGroups = async (
+    user: User,
+    groups: Group[]
+  ): Promise<void> => {
+    for (const g in groups) {
+      const group = groups[g]
+
+      const members = await getGroupMembership(group);
+      const userMembership = R.find(R.pathEq(['user', 'id'], user.id))(members);
+
+      if (userMembership) {
+        try {
+          await keycloakAdminClient.users.delFromGroup({
+            // @ts-ignore
+            id: userMembership.user.id,
+            // @ts-ignore
+            groupId: userMembership.roleSubgroupId
+          });
+        } catch (err) {
+          throw new Error(`Could not remove user from group: ${err.message}`);
+        }
+      }
+
+      const allGroups = await loadAllGroups();
+      const keycloakGroups = await transformKeycloakGroups(allGroups);
+      const data = Buffer.from(JSON.stringify(keycloakGroups)).toString('base64')
+      try {
+        // then attempt to save it to redis
+        await saveRedisKeycloakCache("allgroups", data);
+      } catch (err) {
+        logger.warn(`Couldn't save redis keycloak cache: ${err.message}`);
+      }
```

Comment:
This will run on every group that the user is removed from. This query is quite intense.
Would be better to run it after the user has been removed from all groups

Reply:
Good spot - updated